status/private/tx_history.nim

```diff
@@ -393,114 +519,143 @@ var fetchLock: RLock
+      withRLock(fetchLock):
+        var addresses = initHashSet[string]()
+
+        # Fetch addresses from the channel
+        while chan.peek() > 0:
+
+          let t = chan.recv()
+          let address = t[0]
+          let addOrRemove: bool = t[1]
+          if addOrRemove:
+            addresses.incl(address)
           else:
-            txBlockRange = [dbBlockNumber + 1, lastBlockNumber]
-          var txToData = TransferMap()
-
-          # Only fetch eth transfers if tx count or balance changed
-          if dbTxCount != txCount or dbBalance != balance:
-            fetchEthTxHashes(address, txBlockRange, txToData)
-
-          # ERC-20 logs should be checked anyways
-          fetchErc20Logs(address, txBlockRange, txToData)
-
-          if len(txToData) > 0:
-            fetchTxDetails(address, txToData)
-
-          let transferInfo = TransferInfo(
-            address: address,
-            balance: balance,
-            blockNumber: lastBlockNumber,
-            txCount: txCount)
-          writeToDb(TxDbData(info: transferInfo, txToData: txToData))
-
-    sleep(schedulerInterval)
+            addresses.excl(address)
+
+
+        # Update history for each address
+        for address in items(addresses):
+          var dbData = fetchDbData(address)
+
+          let dbBlockNumber = dbData.info.blockNumber
+          let dbTxCount = dbData.info.txCount
+          let dbBalance = dbData.info.balance
+
+          # If this is an initial wallet scan, all of [dbBlockNumber, dbTxCount, dbBalance]
+          # will be zero
+          var lastBlockNumber = getLastBlockNumber()
+          if lastBlockNumber > dbBlockNumber:
+            # There are new blocks, check if outgoing tx count or balance changed
+            let
+              txCount = getValueForBlock(address, intToHex(lastBlockNumber),
+                RemoteMethod.eth_getTransactionCount)
+              balance = getValueForBlock(address, intToHex(lastBlockNumber),
+                RemoteMethod.eth_getBalance)
+
+
+            var txBlockRange: BlockRange
+            if dbBlockNumber == 0:
+              txBlockRange = txBinarySearch(address, lastBlockNumber)
+            else:
+              txBlockRange = [dbBlockNumber + 1, lastBlockNumber]
+            var txToData = TransferMap()
+
+            # Only fetch eth transfers if tx count or balance changed
+            if dbTxCount != txCount or dbBalance != balance:
+              fetchEthTxHashes(address, txBlockRange, txToData)
+
+            # ERC-20 logs should be checked anyways
+            fetchErc20Logs(address, txBlockRange, txToData)
+
+            if len(txToData) > 0:
+              fetchTxDetails(address, txToData)
+
+            let transferInfo = TransferInfo(
+              address: address,
+              balance: balance,
+              blockNumber: lastBlockNumber,
+              txCount: txCount)
+            writeToDb(TxDbData(info: transferInfo, txToData: txToData))
+
+      sleep(schedulerInterval)
```

Comment:
I realize this code didn't originate with this PR, but we can't `sleep`, as that would block the thread on which nim-status is running. It should be up to a client implementor (e.g. ourselves with the example client) to fetch history in a loop using `sleepAsync`.

Also, as seen later in this module, this logic involves a lock and and an extra thread, which is also a no-go. I'ved created an issue re: this module needing to be overhauled: #266.

Reply:
Definitely agreed. There are no code paths from our public APIs to this module, nor any calls internally to this code. As discussed, we should repurpose it to be consumed by the client in a manner suited to the architecture of the client. In the case of the example chat client, we could start a long-running task that does this work.